packages/connection-storage/src/connection-storage.ts

```diff
@@ -98,11 +107,128 @@ export class ConnectionStorage {
     this.calledOnce = true;
   }
 
-  private static mapStoredConnectionToConnectionInfo(
-    storedConnectionInfo: ConnectionInfo,
+  static async migrateToSafeStorage() {
+    // If user lands on this version of Compass with legacy connections (using storage mixin),
+    // we will ignore those and only migrate connections that have connectionInfo.
+    const connections = (await this.getConnections()).filter(
```

Comment:
If this fails – it could if the connections directory does not exist or is not readable – the entire app wouldn’t start anymore, because this is called early during startup. That does not seem okay.

Reply:
That's a good point. However, we resolved this in [user-data](https://github.com/mongodb-js/compass/blob/eda417439a906b9d70656b8a7020005547febc37/packages/compass-user-data/src/user-data.ts#L89), where as we try to read from a path, we ensure that the directory exists. 

And if the directory is not readable for some reason, we have [ignoreErrors](https://github.com/mongodb-js/compass/blob/main/packages/compass-user-data/src/user-data.ts#L256) prop which defaults to true and this method returns an empty list.

Reply:
> And if the directory is not readable for some reason, we have [ignoreErrors](https://github.com/mongodb-js/compass/blob/main/packages/compass-user-data/src/user-data.ts#L256) 

It doesn't ignore errors from `readdir`, though, which is what I was thinking of

> That's a good point. However, we resolved this in [user-data](https://github.com/mongodb-js/compass/blob/eda417439a906b9d70656b8a7020005547febc37/packages/compass-user-data/src/user-data.ts#L89), where as we try to read from a path, we ensure that the directory exists.

A directory existing doesn't strictly mean that it's readable:

```js
> fs.mkdirSync('/root', {recursive: true})
undefined
> fs.readdirSync('/root')
Uncaught Error: EACCES: permission denied, scandir '/root'
```

I don't think this is something that is likely to happen in Compass, but still, maybe the top-level call to the migration function could be wrapped in a `try/catch` just to make sure?

Reply:
Yes, now that all makes sense - `readdir` being outside try/catch, maybe we should also handle it with `options.ignoreErrors`? 

I'll wrap the migration in try/catch.

Reply:
> maybe we should also handle it with options.ignoreErrors?

Yes, I think that would also make sense :)